### PR TITLE
Sylvia whittle/ 517 swiss army knife

### LINKED ISSUE #517: Remodel entry points and Command Line Interface to use "Swiss Army knife" approach
I think it would be useful to revise the Command Line Interface (CLI) entry point to TopoStats. Currently there are two `run_topostats` and `toposum` but in order to make this extensible I feel we should adopt what is termed the "Swiss Army knife" approach to Command Line Interfaces.

This is what programmes such as `git`, `pre-commit` and many others use. They have a single command for invocation followed by a sub-command. Using `pre-commit` as an example (as its written in Python and provides a good pattern to emulate)...

The main `pre-commit` command has the following help...

```bash
❱ pre-commit --help
usage: pre-commit [-h] [-V]
                  {autoupdate,clean,gc,init-templatedir,install,install-hooks,migrate-config,run,sample-config,try-repo,uninstall,validate-config,validate-manifest,help,hook-impl}
                  ...

positional arguments:
  {autoupdate,clean,gc,init-templatedir,install,install-hooks,migrate-config,run,sample-config,try-repo,uninstall,validate-config,validate-manifest,help,hook-impl}
    autoupdate          Auto-update pre-commit config to the latest repos' versions.
    clean               Clean out pre-commit files.
    gc                  Clean unused cached repos.
    init-templatedir    Install hook script in a directory intended for use with `git config init.templateDir`.
    install             Install the pre-commit script.
    install-hooks       Install hook environments for all environments in the config file. You may find `pre-commit install --install-hooks` more useful.
    migrate-config      Migrate list configuration to new map configuration.
    run                 Run hooks.
    sample-config       Produce a sample .pre-commit-config.yaml file
    try-repo            Try the hooks in a repository, useful for developing new hooks.
    uninstall           Uninstall the pre-commit script.
    validate-config     Validate .pre-commit-config.yaml files
    validate-manifest   Validate .pre-commit-hooks.yaml files
    help                Show help for a specific command.

options:
  -h, --help            show this help message and exit
  -V, --version         show program's version number and exit
```

That is the `positional arguments` are sub-commands, if you want to `run` then you have the following options to that...

```bash
❱ pre-commit run --help
usage: pre-commit run [-h] [--color {auto,always,never}] [-c CONFIG] [--verbose] [--all-files | --files [FILES ...]] [--show-diff-on-failure]
                      [--hook-stage {commit,merge-commit,prepare-commit-msg,commit-msg,post-commit,manual,post-checkout,push,post-merge,post-rewrite}]
                      [--remote-branch REMOTE_BRANCH] [--local-branch LOCAL_BRANCH] [--from-ref FROM_REF] [--to-ref TO_REF] [--commit-msg-filename COMMIT_MSG_FILENAME]
                      [--prepare-commit-message-source PREPARE_COMMIT_MESSAGE_SOURCE] [--commit-object-name COMMIT_OBJECT_NAME] [--remote-name REMOTE_NAME]
                      [--remote-url REMOTE_URL] [--checkout-type CHECKOUT_TYPE] [--is-squash-merge IS_SQUASH_MERGE] [--rewrite-command REWRITE_COMMAND]
                      [hook]

positional arguments:
  hook                  A single hook-id to run

options:
  -h, --help            show this help message and exit
  --color {auto,always,never}
                        Whether to use color in output. Defaults to `auto`.
  -c CONFIG, --config CONFIG
                        Path to alternate config file
  --verbose, -v
  --all-files, -a       Run on all the files in the repo.
  --files [FILES ...]   Specific filenames to run hooks on.
  --show-diff-on-failure
                        When hooks fail, run `git diff` directly afterward.
  --hook-stage {commit,merge-commit,prepare-commit-msg,commit-msg,post-commit,manual,post-checkout,push,post-merge,post-rewrite}
                        The stage during which the hook is fired. One of commit, merge-commit, prepare-commit-msg, commit-msg, post-commit, manual, post-checkout, push, post-
                        merge, post-rewrite
  --remote-branch REMOTE_BRANCH
                        Remote branch ref used by `git push`.
  --local-branch LOCAL_BRANCH
                        Local branch ref used by `git push`.
  --from-ref FROM_REF, --source FROM_REF, -s FROM_REF
                        (for usage with `--to-ref`) -- this option represents the original ref in a `from_ref...to_ref` diff expression. For `pre-push` hooks, this represents the
                        branch you are pushing to. For `post-checkout` hooks, this represents the branch that was previously checked out.
  --to-ref TO_REF, --origin TO_REF, -o TO_REF
                        (for usage with `--from-ref`) -- this option represents the destination ref in a `from_ref...to_ref` diff expression. For `pre-push` hooks, this
                        represents the branch being pushed. For `post-checkout` hooks, this represents the branch that is now checked out.
  --commit-msg-filename COMMIT_MSG_FILENAME
                        Filename to check when running during `commit-msg`
  --prepare-commit-message-source PREPARE_COMMIT_MESSAGE_SOURCE
                        Source of the commit message (typically the second argument to .git/hooks/prepare-commit-msg)
  --commit-object-name COMMIT_OBJECT_NAME
                        Commit object name (typically the third argument to .git/hooks/prepare-commit-msg)
  --remote-name REMOTE_NAME
                        Remote name used by `git push`.
  --remote-url REMOTE_URL
                        Remote url used by `git push`.
  --checkout-type CHECKOUT_TYPE
                        Indicates whether the checkout was a branch checkout (changing branches, flag=1) or a file checkout (retrieving a file from the index, flag=0).
  --is-squash-merge IS_SQUASH_MERGE
                        During a post-merge hook, indicates whether the merge was a squash merge
  --rewrite-command REWRITE_COMMAND
                        During a post-rewrite hook, specifies the command that invoked the rewrite
```

I envisage replacing existing commands with the following (see also table [below](https://github.com/AFM-SPM/TopoStats/issues/517#issuecomment-1517497196) for further thoughts/details)...

| **Current** | **Proposed**  |
|-------------|---------------|
| `run_topostats` | `topostats process` |
| `toposum`       | `topostats summarise` |
| `run_topostats --create-config-file`| `topostats config` |

## Implementation

Following the example of [`pre-commit`](https://github.com/pre-commit/pre-commit/) this would entail introducing a `topostats/main.py` module to provide an entry point of `topostats.main:main`.

[`topostats/main.py`](https://github.com/pre-commit/pre-commit/blob/main/pre_commit/main.py) then imports the various commands from a multitude of sub-modules under [`topostats/commands/*.py`](https://github.com/pre-commit/pre-commit/tree/main/pre_commit/commands) (there is one for each command).

Each arguments are defined for each sub-command within `main.py` in what may be an Abstract Factory design pattern (not quite sure on this front yet!).

## Additional Changes

In addition all documentation (`README.md`, `docs/usage.md` etc.) would also require updating to reflect these changes.

### PR DESCRIPTION
**Pull request by SylviaWhittle**
Closes #517 

This PR changes the entry points for TopoStats. The new way of invoking the processing and summary functions of TopoStats is as such:

```
topostats process <arguments>
```

```
topostats summary <arguments>
```

All previous arguments for `run_topostats` and `toposum` should still function, but as arguments for their respective sub-commands outlined above.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -201,5 +201,4 @@ fixable = ["A", "B", "C", "D", "E", "F", "R", "S", "W", "U"]
 unfixable = []
 
 [project.scripts]
-run_topostats = "topostats.run_topostats:main"
-toposum = "topostats.plotting:main"
+topostats = "topostats.entry_point:entry_point"
```
Comment (ns-rse) [inline pyproject.toml:204]: I think for now we should leave the existing entry points in place and have this as an additional entry point that we can develop and add to. This avoids having to switch whole-sale but allows testing by users who are inclined to do so and allows us to add in the modular processing as we go through refactoring.

**Review by ns-rse — COMMENTED** [review]
I've been giving this some thought in light of the increasing flexibility for processing being requested in terms or running or re-running individual stages and think that we should keep the existing entry point in place for now but develop this new method of interacting in parallel, then when we are happy everything is working and all modules are covered we can make the switch.

To this end changes to the `bug_report.md`, `docs/*.md` and `notebooks/*.md` would need excluding. The original entry-points (`run_topostats` and `toposum`) would need reinstating.

What do you think to this way of proceeding with developing this modular processing approach @SylviaWhittle ?

**Comment by SylviaWhittle** [comment]
Thanks @ns-rse! I like your plan.

I have attempted to keep the old entry point and have reverted all the documentation. I have also added tests for both the old and new entry points I have probably overlooked something that needs testing though.

File: topostats/run_topostats.py [hunk 2]
```diff
@@ -2,7 +2,6 @@
 
 This provides an entry point for running TopoStats as a command line programme.
 """
-import argparse as arg
```
Comment (ns-rse) [inline topostats/run_topostats.py:5]: This will need reinstating too.

File: topostats/plotting.py [hunk 3]
```diff
@@ -24,44 +23,6 @@
 # pylint: disable=too-many-locals
 
 
-def create_parser() -> arg.ArgumentParser:
```
Comment (ns-rse) [inline topostats/plotting.py:27]: Would need this reinstating too for `toposum` entry point to work.

Comment (ns-rse) [reply topostats/plotting.py:27]: Ignore me, I've no looked at `entry_point.py` and see you've moved all of these there.

File: topostats/run_topostats.py (hunk 2 above)
Comment (ns-rse) [reply topostats/run_topostats.py:5]: No they don't.

File: tests/test_entry_point.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    assert returned_args_dict[expected_arg_name] == expected_arg_value
+
+
+def test_entry_point_create_config_file(tmp_path: Path) -> None:
+    """Test that the entry point is able to produce a default config file when asked to."""
+
+    with pytest.raises(SystemExit):
+        entry_point(manually_provided_args=["process", "--create_config_file", f"{tmp_path}/test_create_config.yaml"])
+
+    assert os.path.isfile(f"{tmp_path}/test_create_config.yaml")
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/test_entry_point.py]: Done 👍 thank you! I forgot about this

**Review by ns-rse — APPROVED** [review]
I think this is ready to merge and worth getting into `main` so that we can start extending and modularising the workflow.

**Pull request merged**